add style guide file

a-mccarthy opened the pull request:
Signed-off-by: Abigail McCarthy <mabigail@vmware.com>

closes #2600

a-mccarthy commented:
Spent some time researching why the markdown isn't rendering inside the table, and it turns out this is not supported by the markdown interpreter in use, Redcarpet. Do we know why redcarpet was chosen over the jekyll default kramdown?

michmike commented on site/docs/master/style-guide.md:
```diff
@@ -0,0 +1,313 @@
+_This style guide is adapted from the [Kubernetes style guide](https://kubernetes.io/docs/contribute/style/style-guide/)._
+
+This page outlines writing style guidelines for the Velero documentation.
+These are guidelines, not rules. Use your best judgment as you write documentation, and feel free to propose changes to these guidelines. Changes to the style guide are made by the Velero maintainers as a group. To propose a change or addition, create an issue/PR, or add it to the [community meeting agenda](https://hackmd.io/Jq6F5zqZR7S80CeDWUklkA) and attend the meeting to participate in the dis ...
```
our link to the community meeting agenda could change (as evident when we reached max characters last week). better to link to the community page at https://velero.io/community/.

nrb replied:
@michmike Jonas kept the note the same, but copied the old contents into a new note, so the link didn't actually change.

a-mccarthy replied:
Do we think this link is likely to change a lot? If so, it might make sense to point to the community page, so there is only one place to update the link, or create a variable that we can use where ever we need to, and just update in the config file

nrb commented on site/docs/master/style-guide.md:
```diff
@@ -0,0 +1,313 @@
+
+### Avoid statements about the future
+
+Avoid making promises or giving hints about the future. If you need to talk about
```
So far, there have been only beta.

a-mccarthy replied:
ok updated here to say beta

ashish-amarnath commented on site/_data/master-toc.yml:
```diff
@@ -73,6 +73,8 @@ toc:
         url: /code-standards
       - page: Website guidelines
         url: /website-guidelines
+      - page: Style guide
```
suggest calling this `Documentation Style Guide`

carlisia replied:
+1

a-mccarthy replied:
updated to Documentation style guide in the navigation

ashish-amarnath commented on site/docs/master/style-guide.md:
```diff
@@ -0,0 +1,313 @@
+<table caption="Do and Don't - Links" >
+<tr><th>Do</th><th>Don't</th></tr>
+<tr><td>Write hyperlinks that give you context for the content they link to. For example: Certain ports are open on your machines. See <a href="#check-required-ports">Check required ports</a> for more details.</td><td>Use ambiguous terms such as “click here”. For example: Certain ports are open on your machines. See <a href="#check-required-ports">here</a> for more details.</td></tr>
+<tr><td>Write Markdown-style links: `[link text](URL)`. For example: `[community meeting agenda](https://hackmd.io/Jq6F5zqZR7S80CeDWUklkA)` and the output is  [community meeting agenda](https://hackmd.io/Jq6F5zqZR7S80CeDWUklkA).</td><td>Write HTML-style links: `<a href="/media/examples/link-element-example.css" target="_blank">Visit our tutorial!</a>`, or create links that open in new tabs or win ...
```
```suggestion
<tr><td>Write Markdown-style links: `[link text](URL)`. For example: `[community meeting agenda](https://hackmd.io/Jq6F5zqZR7S80CeDWUklkA)` and the output is  [community meeting agenda](https://hackmd.io/Jq6F5zqZR7S80CeDWUklkA).</td><td>Write HTML-style links: `<a href="/media/examples/link-element-example.css" target="_blank">Visit our tutorial!</a>`, or create links that open in new tabs or windows. For example: `[example website](https://example.com){target="_blank"})`</td></tr>
```

nrb replied:
FYI suggestions will not be accepted by our DCO checker

carlisia commented on site/docs/master/style-guide.md:
```diff
@@ -0,0 +1,313 @@
+<tr><td>The kube-scheduler is the default scheduler for Kubernetes.</td><td>kube-scheduler is the default scheduler for Kubernetes.</td></tr>
+</table>
+
+### Use normal style for string and integer field values
```
@a-mccarthy for me I would rather flip and go with what's currently in the "don't" column as the guide. I think it's important to differentiate between what's normal words in a sentence and a value for a field, especially when we have normal-word sounding values (for example, Always).

a-mccarthy replied:
I agree. I flipped the recommendations :)

carlisia commented on site/docs/master/style-guide.md:
```diff
@@ -0,0 +1,313 @@
+<tr><td>Open the `/_data/concepts.yaml` file.</td><td>Open the /\_data/concepts.yaml file.</td></tr>
+</table>
+
+### Use punctuation inside quotes
```
How about parenthesis? Ponctuation should go inside or outside? I'm just taking this opportunity to ask :) we might not need to add this to the guide.

a-mccarthy replied:
For parentheses, punctuation should go outside the ending ), unless the statement inside the () is a full sentence. But also, i think putting things in () can get a little confusing, or skipped entirely by a reader. So for docs, id say try to decrease the amount of use, and consider writing in a way that avoids ()

carlisia commented:
> Do we know why redcarpet was chosen over the jekyll default kramdown?

@jonasrosland I think you are the most qualified to answer.

nrb commented:
> Do we know why redcarpet was chosen over the jekyll default kramdown?

I think the company that initially did the site preferred it, but PR #2625 moved us to kramdown, so we're off redcarpet now.

a-mccarthy commented:
thanks @nrb! i have to update my branch, but this is excellent news! Its much easier to get the tables to render correctly using kramdown (i hope)

a-mccarthy commented on site/docs/master/style-guide.md:
```diff
@@ -0,0 +1,313 @@
+<table caption="Do and Don't - Links" >
+<tr><th>Do</th><th>Don't</th></tr>
+<tr><td>Write hyperlinks that give you context for the content they link to. For example: Certain ports are open on your machines. See <a href="#check-required-ports">Check required ports</a> for more details.</td><td>Use ambiguous terms such as “click here”. For example: Certain ports are open on your machines. See <a href="#check-required-ports">here</a> for more details.</td></tr>
+<tr><td>Write Markdown-style links: `[link text](URL)`. For example: `[community meeting agenda](https://hackmd.io/Jq6F5zqZR7S80CeDWUklkA)` and the output is  [community meeting agenda](https://hackmd.io/Jq6F5zqZR7S80CeDWUklkA).</td><td>Write HTML-style links: `<a href="/media/examples/link-element-example.css" target="_blank">Visit our tutorial!</a>`, or create links that open in new tabs or win ...
```
i updated this to remove the detail about avoiding blank target links and to use markdown style links instead of html links

a-mccarthy commented on site/docs/master/style-guide.md:
```diff
@@ -0,0 +1,313 @@
+
+        kubectl describe pod <pod-name> -n <namespace>
+
+    If the namespace of the pod is `default`, you can omit the '-n' parameter.
```
Do you mean in kubernetes in general? Or something not right with the example, in whats its trying to say about the style guide rule?

carlisia replied:
Oh you know what it is... I was taking `default` to be the actual name. I would remove the backup and say "If the namespace of the pod is the default namespace...".

a-mccarthy commented on site/docs/master/style-guide.md:
```diff
@@ -0,0 +1,313 @@
+
+<table caption="Do and Don't - Headings" >
+<tr><th>Do</th><th>Don't</th></tr>
+<tr><td>Include a title on each page the page or blog post.</td><td>Use include more than one title headings (#) in a page.</td></tr>
```
updated

a-mccarthy commented:
I updated the style guide file so that it will correctly interpret the markdown in the tables, so the examples display correctly. Its not showing in the display preview, bc that is still using redcarpet as the markdown interpreter instead of kramdown (which the site was recently updated to use)

Should i try to get the latest changes into my branch so the preview displays correctly? 

Essentially, if you need to create a table and style table data with markdown, you need to add `markdown="span"` to all `<td>` where the markdown interpreter should be applied. You'd also need to use HTML table markup to make the table, not markdown. Thoughts about this?

cc: @nrb

nrb commented on site/docs/master/style-guide.md:
```diff
@@ -86,7 +86,7 @@ Many readers speak English as a second language. Avoid jargon and idioms to help
 
 Avoid making promises or giving hints about the future. If you need to talk about
-an alpha feature, put the text under a heading that identifies it as alpha
+an beta feature, put the text under a heading that identifies it as beta
```
a beta feature

a-mccarthy replied:
fixed this

nrb commented:
> Should i try to get the latest changes into my branch so the preview displays correctly?

@a-mccarthy Are you viewing the previews at https://deploy-preview-2619--velero.netlify.app/docs/master/style-guide/, or in the GitHub review previews? I think the GitHub review pages will use their own rendered, while the netlify preview will use ours.

I think it's worth incorporating the changes to the site to make sure you're up-to-date and you can use the Markdown tables, though.

nrb commented on site/docs/master/style-guide.md:
```diff
@@ -0,0 +1,317 @@
+<tr><td>Velero</td><td>Velero should always be capitalized.</td></tr>
+<tr><td>VMware</td><td>VMware should always be correctly capitalized.</td></tr>
+<tr><td>On-premises</td><td>On-premises or on-prem rather than on-premise or other variations.</td></tr>
+<tr><td>Backup</td><td>Backup rather than back up, back-up or other variations.</td></tr>
```
We've been using `backup` as the artifact produced by the process, and `back up` as the verb for the process itself.

I'm happy to converge on the same style for both, though, as having the separately formatted words makes it much harder for folks to know which is correct for the context.

a-mccarthy replied:
I think its better to have a single spelling

nrb commented on site/docs/master/style-guide.md:
```diff
@@ -0,0 +1,317 @@
+<tr><td>VMware</td><td>VMware should always be correctly capitalized.</td></tr>
+<tr><td>On-premises</td><td>On-premises or on-prem rather than on-premise or other variations.</td></tr>
+<tr><td>Backup</td><td>Backup rather than back up, back-up or other variations.</td></tr>
+<tr><td>Plugin</td><td>Plugin rather than plug-in or other variations.</td></tr>
```
I _think_ the VMware-official formatting is `plug-in`, but personally I prefer `plugin` :)

carlisia replied:
^ that is correct. But.... we've been using `plugin` since always :)

a-mccarthy replied:
i'd say we are ok using plugins if that is what you are already using :)

nrb commented on site/docs/master/style-guide.md:
```diff
@@ -0,0 +1,317 @@
+
+
+### Line breaks
+Use a single newline to separate block-level content like headings, lists, images, code blocks, and others. The exception is second-level headings, where it should be two newlines. Second-level headings follow the first-level (or the title) without any preceding paragraphs or texts. A two line spacing helps visualize the overall structure of content in a code editor better.
```
This is a personal preference, but I line break at sentences in Markdown to make content easier to review. Curious what others think about that practice; clearly this isn't how it's viewed in the Kubernetes docs team.

carlisia replied:
I think this direction is overkill. Mostly we follow what's already there, and this is only for consumption of whoever is reading the markdown code, not for users of the documentation. I'd remove this one.

a-mccarthy replied:
To nolan's point, the goal of line break rules is to make markdown easier to read and update. Do you mostly review docs in Github or using the netlify preview? To me this rule seems like too much, but if the team felt strongly about having character limits and line breaks done a certain way we can change it to something else

nrb replied:
I do both - I like GitHub's diff view for seeing what changed, and netlify's preview for how it renders.

I am fine with dropping the line break guideline entirely.

a-mccarthy replied:
Ok removed the line break rule

nrb commented on site/docs/master/style-guide.md:
```diff
@@ -0,0 +1,317 @@
+<tr><td markdown="span">Use single backticks to enclose inline code. For example, `var example = true`.</td><td>Use two asterisks (`**`) or an underscore (`_`) to enclose inline code. For example, **var example = true**.</td></tr>
+<tr><td>Use triple backticks (\`\`\`) before and after a multi-line block of code for fenced code blocks.</td><td>Use multi-line blocks of code to create diagrams, flowcharts, or other illustrations.</td></tr>
+<tr><td>Use meaningful variable names that have a context.</td><td>Use variable names such as 'foo','bar', and 'baz' that are not meaningful and lack context.</td></tr>
+<tr><td>Remove trailing spaces in the code.</td><td>Add trailing spaces in the code, where these are important, because the screen reader will read out the spaces as well.</td></tr>
```
Should this say `a screen reader`?

a-mccarthy replied:
yes! updated

a-mccarthy commented:
>Are you viewing the previews at https://deploy-preview-2619--velero.netlify.app/docs/master/style-guide/, or in the GitHub review previews? I think the GitHub review pages will use their own rendered, while the netlify preview will use ours.

@nrb I am looking that the netlify preview. How do I update the PR branch to include the latest changes from the master branch?

I was able to update my branch locally with the new markdown by pulling in the latest changes from master branch. And things looked correct with the new markdown. But I'm not sure how to get that into the PR so the netlify preview works correctly. when i tried to push to github, i got an error saying i didn't have permission to update certain files

a-mccarthy commented:
I was able to update my branch with the latest changes from master, thanks to Jonas. The deploy preview is now showing the correctly displaying style guide tables. 

@nrb i addressed your typo changes today as well, can you review again?

a-mccarthy commented:
Updated with allowlist and denylist to the terms list. 

@nrb I addressed your requested changes, is that something you can clear with another review? I'm still learning Github reviews

nrb commented:
@a-mccarthy Review flipped, thanks!
